src/main/cpp/src/json_parser.hpp

```diff
@@ -0,0 +1,1786 @@
+ * limitation.
+ *
+ * e.g.: The length of number -123.45e-67 is 7. if maximum JSON number length
+ * is 6, then this number is a invalid number.
+ */
+constexpr int curr_max_num_len = 1000;
+
+/**
+ * whether allow tailing useless sub-string in JSON.
+ *
+ * If true, e.g., the following invalid JSON is allowed, because prefix {'k' :
+ * 'v'} is valid.
+ *   {'k' : 'v'}_extra_tail_sub_string
+ */
+constexpr bool curr_allow_tailing_sub_string = true;
+
+/**
+ * JSON token enum
+ */
+enum class json_token {
+  // start token
+  INIT = 0,
+
+  // successfully parsed the whole JSON string
+  SUCCESS,
+
+  // get error when parsing JSON string
+  ERROR,
+
+  // '{'
+  START_OBJECT,
+
+  // '}'
+  END_OBJECT,
+
+  // '['
+  START_ARRAY,
+
+  // ']'
+  END_ARRAY,
+
+  // e.g.: key1 in {"key1" : "value1"}
+  FIELD_NAME,
+
+  // e.g.: value1 in {"key1" : "value1"}
+  VALUE_STRING,
+
+  // e.g.: 123 in {"key1" : 123}
+  VALUE_NUMBER_INT,
+
+  // e.g.: 1.25 in {"key1" : 1.25}
+  VALUE_NUMBER_FLOAT,
+
+  // e.g.: true in {"key1" : true}
+  VALUE_TRUE,
+
+  // e.g.: false in {"key1" : false}
+  VALUE_FALSE,
+
+  // e.g.: null in {"key1" : null}
+  VALUE_NULL
+
+};
+
+/**
+ * JSON parser, provides token by token parsing.
+ * Follow Jackson JSON format by default.
+ *
+ *
+ * For JSON format:
+ * Refer to https://www.json.org/json-en.html.
+ *
+ * Note: when setting `allow_single_quotes` or `allow_unescaped_control_chars`,
+ * then JSON format is not conventional.
+ *
+ * White space can only be 4 chars: ' ', '\n', '\r', '\t',
+ * Jackson does not allow other control chars as white spaces.
+ *
+ * Valid number examples:
+ *   0, 102, -0, -102, 0.3, -0.3
+ *   1e-5, 1E+5, 1e0, 1E0, 1.3e5
+ *   1e01 : allow leading zeor after 'e'
+ *
+ * Invalid number examples:
+ *   00, -00   Leading zeroes not allowed
+ *   infinity, +infinity, -infinity
+ *   1e, 1e+, 1e-, -1., 1.
+ *
+ * When `allow_single_quotes` is true:
+ *   Valid string examples:
+ *     "\'" , "\"" ,  '\'' , '\"' , '"' , "'"
+ *
+ *  When `allow_single_quotes` is false:
+ *   Invalid string examples:
+ *     "\'"
+ *
+ *  When `allow_unescaped_control_chars` is true:
+ *    Valid string: "asscii_control_chars"
+ *      here `asscii_control_chars` represents control chars which in Ascii code
+ * range: [0, 32)
+ *
+ *  When `allow_unescaped_control_chars` is false:
+ *    Invalid string: "asscii_control_chars"
+ *      here `asscii_control_chars` represents control chars which in Ascii code
+ * range: [0, 32)
+ *
+ */
+template <bool allow_single_quotes           = curr_allow_single_quotes,
+          bool allow_unescaped_control_chars = curr_allow_unescaped_control_chars,
+          int max_json_nesting_depth         = curr_max_json_nesting_depth,
+          int max_string_utf8_bytes          = curr_max_string_utf8_bytes,
+          int max_num_len                    = curr_max_num_len,
+          bool allow_tailing_sub_string      = curr_allow_tailing_sub_string>
```

Comment:
Nit: Oh no, too many template parameters. They will make compilation very slow. We should better use run time parameters, except only the parameters that need to be known at compile time. In future work we have to clean up this.

Reply:
Only have one template combination.

Reply:
Will do in a follow-on.